sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,974 @@
+
+[_Note:_ The {dpcpp} implementation of this extension currently has the
+restrictions listed in this section.
+In the future, restrictions tied to the integration header approach might be
+formalized in the core SYCL specification and tied to a macro, similar to the
+feature set macros that exist already.
+_{endnote}_]
+
+Implementations of SYCL that use the integration header technique have
+additional restrictions for functions that are declared as free function
+kernels.
+These implementations automatically insert forward declarations of the free
+function kernels at the top of the translation unit.
+This has ramifications on how the application may declare the free function
+kernels, on the types that may be used in those declarations, and on the way
+the application may reference these kernel identifiers.
+The following example illustrates the forward declarations that the
+implementation inserts:
+
+```
+// Forward declarations of types used by the kernel functions.
+struct mystruct;
+enum myenum : int;
+
+// Each kernel is forward declared in the same namespace in which the
+// application declares it.
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
+void kernel1(int *);
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
+void kernel2(mystruct, myenum);
+
+template<typename T>
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
+void kernel3(T *);
+
+namespace ns {
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
+void kernel4(int *);
+}
+```
+
+(The lines using `SYCL_EXT_ONEAPI_FUNCTION_PROPERTY` are exposition-only.
+Implementations will probably emit some implementation-specific code here
+instead of using the macro because the macro and the `range_kernel` property
+are probably defined in the `<sycl/sycl.hpp>` header, which does not get
+get included until after the integration header.)
+
+As a result, these implementations impose additional restrictions for functions
+that are declared as free function kernels:
+
+* The function must be declared at namespace scope.
+
+* Any type used in the declaration of a parameter must be one of the allowed
+  types listed below.
+
+* If the function is instantiated from a template, any type used to instantiate
+  the template must be one of the allowed types listed below.
+
+* Uses of function identifiers in the application must assume that the free
+  function kernels are forward declared at the top of the translation unit.
+  Note that this can also affect references to functions that are not declared
+  as free functions kernels as illustrated below.
++
+```
+void foo(int) {/*...*/}
+
+void caller() {
+  auto *pf = foo;  // This is ambiguous because foo(float) is forward declared
+                   // in the integration header
+}
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
+void foo(float) {/*...*/}
+```
+
+The allowed types are:
+
+* A {cpp} fundamental type.
+* A class or struct that is defined at namespace scope.
```

Comment:
@gmlueck can you clarify this requirement? What does it mean defined at namespace scope?
The structure/class needs to be defined in the same namespace as a free function kernel?

Reply:
Namespace scope is sometimes referred to as "file scope".  For example, this restriction prohibits using a struct/class type that is defined inside a function or inside another struct or class.  Here are some examples to illustrate:

```
struct foo1;  // defined at namespace scope

namespace bar {
struct foo2;  // defined at namespace scope
}

namespace {
struct foo3;  // defined at namespace scope
}

void func() {
  struct foo4;  // NOT defined at namespace scope (this is block scope)
}

struct outer {
  struct foo5;  // NOT defined at namespace scope (this is class scope)
};
```

> The structure/class needs to be defined in the same namespace as a free function kernel?

No, this is not necessary.  The struct/class can be defined in a different namespace than the function.  For example, this is OK:

```
namespace ns1 {
struct mystruct;
}

namespace ns2 {

SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
void iota(ns1::mystruct *p) { /*...*/  }

}
```

Reply:
Thank you for clarification.
What if the structure is defined at the file scope in another file? i.e. shared header?
According to what you wrote, it should be fine

Reply:
Yes, that's fine.  There are NO limitations on how you split your code between the main *.cpp file and the header files it includes.